### PR TITLE
[cpp] Generic Triplet Margin Loss

### LINKED ISSUE #43342: Proposal: Generic Triplet-Margin Loss
## 🚀 Feature
We'd like to add a distance-agnostic version of the [TripletMarginLoss](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), as the current version only supports l_p norms.  This was something we implemented as part of a recent investigation for embedding learning and thought could be helpful to the PyTorch userbase at large.

## Motivation

1. Extensibility: decoupling the distance function and the loss computation allow for more flexibility for users.  PyTorch currently has a [CosineEmbeddingLoss](https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html), but that serves a somewhat different purpose and doesn't really work for users wanting a triplet-margin loss with cosine distance.

2. Existing use cases: several papers have proposed triplet loss functions with cosine distance ([1](https://www.isca-speech.org/archive/Interspeech_2018/pdfs/1209.pdf), [2](https://www.researchgate.net/publication/316736728_Deep_Speaker_an_End-to-End_Neural_Speaker_Embedding_System#read)) or have generally used cosine-based metrics ([1](https://openaccess.thecvf.com/content_cvpr_2018/papers/Wang_CosFace_Large_Margin_CVPR_2018_paper.pdf), [2](https://arxiv.org/pdf/1812.00442.pdf)).  [PyTorch-BigGraph](https://arxiv.org/abs/1903.12287) also does something similar with its ranking loss.

3.  Products like Tensorflow [decouple](https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/triplet.py) the distance functions and even allow for custom distance metrics

## Pitch

We have a working implementation of this (see screenshots) and propose to use it directly, with adjustments to fit the existing PyTorch code structure.  We'll tentatively call these `torch.nn.GenericTripletMarginLoss` and `F.generic_triplet_margin_loss`.  In the proposed implementation, `GenericTripletMarginLoss` will be a stateful wrapper for `F.generic_triplet_margin_loss`, and the loss computation will occur in the latter.

The module signatures will resemble that of the [production module](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), with an analogous signature for the [functional module](https://pytorch.org/docs/master/_modules/torch/nn/modules/loss.html#TripletMarginLoss).  The removed parameters (`p` and `eps`) are specific to l_p norm computation; `size_average` and `reduce` are removed because of deprecation.  There are two new parameters: `distance_function` and `is_similarity_function`.  `distance_function` is a function/module $$f: R^{n \cross d} \rightarrow R^n$$ that gives an index-wise distance metric between two tensors of embeddings.  This can be, for example, `torch.nn.PairwiseDistance` or `torch.nn.CosineSimilarity`.  `is_similarity_function` is a boolean that denotes whether `distance_function` is a distance or similarity function: this will indicate whether to flip the signs in distance computation.

## Alternatives

Beyond the straightforward alternative of leaving this to the user to implement every time, we haven't come up with any other alternatives.

One open question is whether the Python implementation is performant enough, or if this would need to be added in aTen as well.

## Additional context

image

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by ethch18**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43639 [cpp] Generic Triplet Margin Loss**

Differential Revision: [D23351411](https://our.internmc.facebook.com/intern/diff/D23351411)

[WIP; tests are forthcoming]

This pull request resolves #43342 by refactoring the C++ and Python implementations to take an additional function argument.  The high-level design is as follows:

* Add an optional `distance_function` argument of type `std::function<Tensor(const Tensor&, const Tensor&)>` to the C++ functional and module APIs
* [TODO] Add an optional `distance_function` argument to the Python functional and module APIs
* Add a `is_similarity_function` boolean argument (default = `False`) to the C++ functional and module APIs
* Refactor the aTen implementation of `torch::triplet_margin_loss` to take/use the new `distance_function` and `is_similarity_function` arguments.

Currently blocked on figuring out how to update the signature in `native_functions.yaml` to accept the `std::function` arg; workarounds or alternatives welcomed.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3158,7 +3158,7 @@
 - func: _trilinear(Tensor i1, Tensor i2, Tensor i3, int[] expand1, int[] expand2, int[] expand3, int[] sumdim, int unroll_dim=1) -> Tensor
   use_c10_dispatcher: full
 
-- func: triplet_margin_loss(Tensor anchor, Tensor positive, Tensor negative, float margin=1.0, float p=2, float eps=1e-06, bool swap=False, int reduction=Mean) -> Tensor
+- func: triplet_margin_loss(Tensor anchor, Tensor positive, Tensor negative, float margin=1.0, float p=2, float eps=1e-06, bool swap=False, int reduction=Mean, TODO? distance_function=None, bool is_similarity_function=False) -> Tensor
```
Comment (ethch18) [inline aten/src/ATen/native/native_functions.yaml:3161]: Not sure what to put here, as it seems that native_functions doesn't support function args.  Or is there an alternate place to declare functions that would be C++ only?

**Comment by ethch18** [comment]
Closing after offline discussion, as this was a draft
